File: workspace/A4-Python/CustomerPrQ.py
'''
*******************************************************
Crowd Organizer App
Author: Caleb Viola
CustomerPrQ.py
Implementation of min-heap for managing priority queue.
*******************************************************
'''
import math


class CustomerPrQ():
    #************************
    def __init__(self, name):
        self.initialPriorityValue = 101
        self.heapNodes = []  # Heap storage
        self.fileName = name

    #*******************************
    def arrangeCustomerQ(self, log):
        log.write('STORE IS OPENING\n')
        with open(self.fileName, 'r') as myfile:
            for line in myfile:
                if not line.startswith('//', 0, 2):
                    self.addCustomerToQ(line.rstrip('\n').split(','), log)
        log.write('>>> Initial heap built containing {} nodes\n\n'.format(len(self.heapNodes)))

    #**************************************
    def serveRemainingCustomers(self, log):
        log.write('\n')
        log.write('STORE IS CLOSING\n')
        log.write('>>> Heap currently has {} nodes remaining\n'.format(len(self.heapNodes)))
        while self.heapNodes:
            self.serveACustomer(log)
        log.write('>>> Heap is now empty\n')
        log.write('>>> Program terminating\n')

    '''
    :args contains = (raw string, toPrint boolean, log object)
    '''
    #*******************************************
    def addCustomerToQ(self, *args):
        raw, toPrint, log = args if (len(args) > 1) else args[0]
        nextInLine = self.__determinePriorityValue(raw[1:])
        if toPrint:
            log.write('ADDING: {} ({})\n'.format(''.join(raw[:1]), nextInLine))
        self.__heapInsert(''.join(raw[:1]), nextInLine)
        self.initialPriorityValue += 1

    #*****************************
    def serveACustomer(self, log):
        if self.heapNodes:
            log.write('SERVING: {} ({})\n'.format(self.heapNodes[0].name,
                                                  self.heapNodes[0].priorityValue))
            self.__heapDelete()
        else:
            log.write('>>> Heap is empty\n')

    #*******************************************
    def __heapInsert(self, name, priorityValue):
        x = Node(name, priorityValue)
        self.heapNodes.append(x)
        self.walkUp(len(self.heapNodes) - 1)

    #**********************
    def __heapDelete(self):
        minItem = self.heapNodes[0]
        self.heapNodes[0] = self.heapNodes[len(self.heapNodes) - 1]
        self.heapNodes.pop()
        self.walkDown(0)
        return minItem

    #***************************************
    def __determinePriorityValue(self, raw):
        subtract = 0
        for jumpTheQPoints in raw:
            if not jumpTheQPoints == '':
                subtract += {jumpTheQPoints: (15 if jumpTheQPoints.isnumeric() and (int(jumpTheQPoints) >= 65 or int(jumpTheQPoints) >= 80) else 0),
                             'employee': 25,
                             'owner': 80,
                             'vip': 5,
                             'superVIP': 10}[jumpTheQPoints]
        return self.initialPriorityValue - subtract


    #***************************
    def walkUp(self, startFrom):
        i = startFrom
        while (i > 0) and \
                (self.heapNodes[i].priorityValue < self.heapNodes[math.trunc((i - 1) / 2)].priorityValue):
            self.heapNodes[i], self.heapNodes[math.trunc((i - 1) / 2)] = \
                self.heapNodes[math.trunc((i - 1) / 2)], self.heapNodes[i]  # Swap
            i = math.trunc((i - 1) / 2)

    #*****************************
    def walkDown(self, startFrom):
        i = startFrom
        smCh = self.subOfSmCh(i)
        while (2 * i + 1) <= len(self.heapNodes) - 1 and \
                (self.heapNodes[i].priorityValue > self.heapNodes[smCh].priorityValue):
            self.heapNodes[i], self.heapNodes[smCh] = \
                self.heapNodes[smCh], self.heapNodes[i]  # Swap
            i = smCh
            smCh = self.subOfSmCh(i)

    #**********************
    def subOfSmCh(self, i):
        if ((2*i + 2) > len(self.heapNodes) - 1 or
                (self.heapNodes[2*i + 1].priorityValue <= self.heapNodes[2*i + 2].priorityValue)):
            return 2*i + 1
        else:
            return 2*i + 2


class Node:
    #***************************************
    def __init__(self, name, priorityValue):
        self.name = name
        self.priorityValue = priorityValue

File: workspace/CrowdOrganizer/src/CustomerPrQ.java
/**
* PROJECT: CrowdOrganizer     CLASS: CustomerPrQ
* AUTHOR: Benjamin Johnson
* FILES ACCESSED: LineAt6Am.csv (or what ever file name is passed into the setupPrQ
* 	 method)
* INTERNAL INDEX STRUCTURE: A minimum heap stored in a array of heapNodes             
* FILE STRUCTURE: 
* DESCRIPTION: This Class will create a minimum heap using the heap insert method, 
* 	it also has the appropriate methods to insert and delete from the heap using walk
* 	up and walk down to correct issues.
***********************************************************************************/

import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.util.Scanner;

public class CustomerPrQ
{
	//**************************** PRIVATE DECLARATIONS ************************
	private BufferedWriter log;
	private int N;
	private final int maxN=50;
	private HeapNode[] heap;
	private int initialPriorityValue = 100;
	private int priorityValue;
	
	//**************************** PUBLIC CONSTRUCTOR **************************
	public CustomerPrQ(BufferedWriter log)
	{
		this.log = log;
		heap = new HeapNode[maxN];
		N = 0;
		
	}
	//**************************** PUBLIC SERVICE METHODS **********************
	
	/****************************** emptyPrq ***************************/
	public void emptyPrq() throws IOException
	{
		log.write(String.format("**heap currently has %d nodes remaining \n", N));
		
		while(!heapEmpty())
		{
			log.write("SERVE CUSTOMER: ");
			heapDelete();
		}
		log.write("**heap now empty \n");
				
	}
	
	/****************************** insertNewCustomer ***************************/
	public void insertNewCustomer(String[] info) throws IOException
	{
		String[] points = new String[4];
		
		for(int i=0;i<4;i++)
		{
			points[i] = info[i+1];
		}
		
		determinePriorityValue(points);
		log.write(String.format("%s (%d) \n", points[0],priorityValue));
		heapInsert(points[0], priorityValue);
		
	}
	public void serveAcustomer() throws IOException
	{
		heapDelete();
	}
	
	/****************************** setupPrQ ***************************/
	public void setupPrQ(String filename) throws IOException
	{
		
		File file = new File(filename);
		Scanner input = new Scanner(file);
		
		log.write(String.format("**Opened file %s \n", filename));
		
		String line;
		String[] info;
		
		while(input.hasNextLine())
		{
			line = input.nextLine();
			
			if(!line.startsWith("//"))
			{
				info = line.split(",");
				
				determinePriorityValue(info);
				
				heapInsert(info[0], priorityValue);
			}
		}
		
		log.write(String.format("**initial heap built containing %d nodes \n", N));
		log.write(String.format("**Closed file %s \n", filename));
		
		input.close();	
	}
	//**************************** PRIVATE METHODS *****************************
	
	/****************************** heapDelete *********************************/
	private void heapDelete() throws IOException
	{
		log.write(String.format("%s (%d) \n",heap[0].getName(),heap[0].getPriorityValue()));
		heap[0] = heap[N-1];	
		N--;
		
		walkDown(0);	
	}
	
	/****************************** heapInsert ***************************/
	private void heapInsert(String value, int pv)
	{
		HeapNode newNode = new HeapNode(value, pv);
		
		heap[N] = newNode;
		N++;
		
		walkUp(N-1);
		
	}
	
	/****************************** walkDown ***************************/
	private void walkDown(int i)
	{
		int smallestChild = getSmallestChild(i);
		
		HeapNode temp = heap[i];
		
		while((2*i+1) <= (N-1) && (temp.getPriorityValue() > heap[smallestChild].getPriorityValue()))
		{
			heap[i] = heap[smallestChild];
			
			i = smallestChild;
			
			smallestChild = getSmallestChild(i);
		}
		heap[i] = temp;		
		
	}

	/****************************** walkUp ***************************/
	private void walkUp(int i)
	{
		
		HeapNode temp=heap[i];
		while(i > 0 && temp.getPriorityValue() < heap[(i-1)/2].getPriorityValue())
		{
		
				heap[i] = heap[(i-1)/2];
				i = (i-1)/2;
			
		}
		heap[i] = temp;
	}
	
	/****************************** determinePriorityValue **********************/
	private void determinePriorityValue(String[] points) throws IOException
	{
		int age = new Integer(points[3]);
		priorityValue = ++initialPriorityValue;
		
		
		for(int i = 1; i<3;i++)
		{
			switch (points[i].toLowerCase()) 
			{
				case "employee":
					priorityValue-=25;
					
					break;
				case "owner":
					priorityValue-=80;
					break;
				case "vip":
					priorityValue-=5;
					break;
				case "supervip":
					priorityValue-=10;
					break;
				
			}
		}
		if(age >= 65)
		{
			priorityValue-= 15;
			
		}
		
		if(age >=80)
		{
			priorityValue-= 15;
		}

	}
	
	/****************************** heapEmpty ***************************/
	private boolean heapEmpty()
	{
		if(N==0)
		{
			return true;
		}
		else
		{
			return false;
		}
	}
	
	/****************************** getSmallestChild ***************************/
	private int getSmallestChild(int i)
	{
		if((2*i+2) > (N-1) || heap[2*i+1].getPriorityValue() <= heap[2*i+2].getPriorityValue())
		{
			return (2*i)+1;
		}
		else
		{
			return (2*i)+2;
		}
	}
}
